OHHTTPStubs/UnitTests/Test Suites/NSURLSessionTests.m

```diff
@@ -240,13 +271,102 @@ - (void)test_NSURLSessionDefaultConfig_notFollowingRedirects
         NSURLSessionConfiguration* config = [NSURLSessionConfiguration defaultSessionConfiguration];
         NSURLSession *session = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:nil];
 
-        NSDictionary* json = @{@"Success": @"Yes"};
-        [self _test_redirect_NSURLSession:session jsonForStub:json completion:^(NSError *errorResponse, NSHTTPURLResponse *redirectResponse, id jsonResponse) {
+        [self _test_redirect_NSURLSession:session httpMethod:@"GET" jsonBody:nil delays:0.1 redirectStatusCode:301
+                               completion:^(NSString *redirectedRequestMethod, id redirectedRequestJSONBody, NSHTTPURLResponse *redirectHTTPResponse, id finalJSONResponse, NSError *errorResponse)
+        {
+            XCTAssertNil(redirectedRequestMethod, @"Expected no redirected request to fire");
+            XCTAssertNil(redirectedRequestJSONBody, @"Expected no redirected request to fire");
+            XCTAssertNotNil(redirectHTTPResponse, @"Redirect response should have been received");
+            XCTAssertNil(finalJSONResponse, @"Unexpected data received when no redirect");
             XCTAssertNil(errorResponse, @"Unexpected error");
-            XCTAssertNotNil(redirectResponse, @"Redirect response should have been received");
-            XCTAssertEqual(301, [redirectResponse statusCode], @"Expected 301 redirect");
-            XCTAssertNil(jsonResponse, @"Unexpected data received");
         }];
+    }
+    else
+    {
+        NSLog(@"/!\\ Test skipped because the NSURLSession class is not available on this OS version. Run the tests a target with a more recent OS.\n");
+    }
+}
+
+/**
+ Verify that redirects of different methods and status codes are handled properly and
+ that we retain the HTTP Method for specific HTTP status codes as well as the data payload.
+ **/
+- (void)test_NSURLSessionDefaultConfig_MethodAndDataRetentionOnRedirect
+{
+    _shouldFollowRedirects = YES;
+
+    if ([NSURLSessionConfiguration class] && [NSURLSession class])
+    {
+        NSURLSessionConfiguration* config = [NSURLSessionConfiguration defaultSessionConfiguration];
+        NSURLSession *session = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:nil];
+
+        NSDictionary* json = @{ @"query": @"Hello World" };
+        NSArray<NSString*>* allMethods = @[@"GET", @"HEAD", @"POST", @"PATCH", @"PUT"];
+
+        /** 301: GET, HEAD, POST, PATCH, PUT **/
+        for (NSString* method in allMethods) {
+            [self _test_redirect_NSURLSession:session httpMethod:method jsonBody:json delays:0.0 redirectStatusCode:301
+                                   completion:^(NSString *redirectedRequestMethod, id redirectedRequestJSONBody, NSHTTPURLResponse *redirectHTTPResponse, id finalJSONResponse, NSError *errorResponse)
+            {
+                XCTAssertEqualObjects(redirectedRequestMethod, method, @"Expected the method to be the same as was sent");
```

Comment:
Maybe it's me but, I find the message of this assert not clear. Something more like "Expected the method to be unchanged"?

Reply:
Good point, fixed.